WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift

```diff
@@ -49,15 +48,143 @@ struct UpgradesView: View {
                 EmptyWaitingView()
             case .completed:
                 EmptyCompletedView()
-            default:
-                EmptyView()
+            case .error(let upgradeError):
+                UpgradesErrorView(upgradeError,
+                                  onRetryButtonTapped: {},
+                                  onRetryPaymentButtonTapped: {},
+                                  onCancelUpgradeTapped: {})
+                .padding(Layout.padding)
             }
+            Spacer()
         }
         .navigationBarTitle(UpgradesView.Localization.navigationTitle)
         .padding(.top)
+        .background(Color(.listBackground).ignoresSafeArea())
     }
 }
 
+struct UpgradesErrorView: View {
+
+    let upgradeError: UpgradesError
+    let errorTitle: String? = nil
+    let errorSubtitle: String? = nil
+    let shouldRetry: Bool = true
+
+    /// Closure invoked when the "Retry" button is tapped
+    var onRetryButtonTapped: (() -> Void) = {}
+
+    /// Closure invoked when the "Try payment again" buton is tapped
+    var onRetryPaymentButtonTapped: (() -> Void) = {}
+
+    /// Closure invoked when the "Cancel upgrade" button is tapped
+    var onCancelUpgradeTapped: (() -> Void) = {}
+
+    init(_ upgradeError: UpgradesError,
+         onRetryButtonTapped: @escaping (() -> Void),
+         onRetryPaymentButtonTapped: @escaping (() -> Void),
+         onCancelUpgradeTapped: @escaping (() -> Void) ) {
+        self.upgradeError = upgradeError
+        self.onRetryButtonTapped = onRetryButtonTapped
+        self.onRetryPaymentButtonTapped = onRetryPaymentButtonTapped
+        self.onCancelUpgradeTapped = onCancelUpgradeTapped
+    }
+
+    var body: some View {
+        VStack(alignment: .center) {
+            Image(uiImage: .planUpgradeError)
```

Comment:
There's no benefit making a UIImage here – just use the Image with the filename, again it works better in previews and has less indirection. If you do it that way and the file doesn't exist, it'll show a placeholder instead, so there's no optionality to deal with.

Reply:
Gotcha! Changed here: 73f03156357dcb8b3a65b57ec83e4f83f473029c